packages/office-ui-fabric-react/src/components/ScreenReaderAlert/ScreenReaderAlert.tsx

```diff
@@ -0,0 +1,163 @@
+   * and browser combinations.
+   */
+  private _renderIndex: number = 0;
+
+  /**
+   * The non-empty string which represents the latest string readout by screen readers.
+   *
+   * The reason of having this variable is that we have a timeout to clear the text in the live region.
+   * We shouldn't let screen readers read the same string if the indicator is not changed. Use this to let React
+   * render the live region only when text passed from props is changed or the indicator is changed.
+   */
+  private _previousValidString: string = '';
+
+  private _clearTextTimeout: number;
+
+  constructor(props: IScreenReaderAlertProps) {
+    super(props);
+
+    this.state = {
+      alertText: this._getTextFromProps(React.Children.toArray(props.children))
+    };
+  }
+
+  public componentWillReceiveProps(nextProps: IScreenReaderAlertProps): void {
+    if (this._clearTextTimeout) {
+      // Clear the timeout of removing text to avoid the new text being removed.
+      window.clearTimeout(this._clearTextTimeout);
+      this._clearTextTimeout = undefined;
+    }
+
+    this.setState({
+      alertText: this._getTextFromProps(React.Children.toArray(nextProps.children))
+    });
+  }
+
+  public shouldComponentUpdate(nextProps: IScreenReaderAlertProps, nextState: IScreenReaderAlertState): boolean {
+    // When indicator is not changed, do NOT need to render the string if it's same with the previous non-empty string.
+    const isValidStringChanged: boolean = nextState.alertText !== this._previousValidString;
+
+    return (this.props.indicator !== nextProps.indicator || isValidStringChanged);
+  }
+
+  public componentDidUpdate(nextProps: IScreenReaderAlertProps): void {
+    this._clearTextByTimeout();
+    this._updateValidString();
+  }
+
+  public componentDidMount(): void {
+    this._clearTextByTimeout();
+    this._updateValidString();
+  }
+
+  public render(): JSX.Element {
+    return (
+      <div className={ styles.screenReaderAlert } key={ this._renderIndex++ }>
+        {
+          this.props.readingMode !== ReadingMode.DoNotRead && this.state.alertText && (
+            <p
+              role={ this._role }
+              aria-live={ this._ariaLive }
+              aria-atomic={ true }>
+              { this.state.alertText }
+            </p>
+          )
+        }
+      </div>
+    );
+  }
+
+  private get _role(): string {
+    return this.props.readingMode === ReadingMode.ReadImmediately ? 'alert' : undefined;
+  }
+
+  private get _ariaLive(): string {
+    switch (this.props.readingMode) {
+      case ReadingMode.DoNotRead:
+        return 'off';
+      case ReadingMode.ReadAfterOtherContent:
+        return 'polite';
+      case ReadingMode.ReadImmediately:
+        return 'assertive';
+      default:
+        return undefined;
+    }
+  }
+
+  private _getTextFromProps(root: React.ReactChild | React.ReactChild[]): string {
+    let text: string = '';
```

Comment:
Yeah it'll get removed after I address your another comment to move the text to props instead of children